src/infoview.ts

```diff
@@ -1,185 +1,216 @@
-export class InfoProvider implements TextDocumentContentProvider, Disposable {
-    scheme = 'lean-info';
+h1
+{
+  display: block;
+  width: 100%;
 
-    documentSelector: DocumentSelector = {
-        scheme: this.scheme,
-    };
+  margin-bottom: 1ex;
+  margin-top: 1ex;
+  border-bottom: 1px solid black;
+  padding: 0px;
 
-    private leanInfoPath = '.lean-info';
-    leanInfoUrl = Uri.parse(`${this.scheme}:${this.leanInfoPath}`);
+  font-weight: normal;
+  font-size: 100%;
+}
 
-    private changedEmitter = new EventEmitter<Uri>();
-    onDidChange = this.changedEmitter.event;
-    private documents = new Map<string, InfoDocument>();
-    private subscriptions: Disposable[] = [];
+div.error h1 {
+  color: red;
+}
 
-    constructor(private server: Server, private leanDocs: DocumentSelector) {
-        this.subscriptions.push(workspace.onDidCloseTextDocument((doc) => {
-            const uri = doc.uri.toString();
-            if (this.documents.has(uri)) {
-                this.documents.get(uri).dispose();
-                this.documents.delete(uri);
-            }
-        }));
+div.warning h1 {
+  color: darkorange;
+}
+
+div.information h1 {
+  color: darkgreen;
+}
+
+#messages div {
+  border-top: 5px solid transparent;
+}
+
+.vscode-light h1 {
+  border-bottom: 1px solid black;
+}
+
+.vscode-dark h1 {
+  border-bottom: 1px solid white;
+}
+
+.vscode-dark div.highlight {
+  background: darkolivegreen;
+}
+
+            </style>
+            </head>
+            <body>` +
+                this.renderGoal() +
+                `<div id="messages">` + this.renderMessages() + `</div>`+
+            `</body></html>`;
     }
 
-    dispose() {
-        this.documents.forEach((v) => v.dispose());
-        for (const s of this.subscriptions) s.dispose();
+    private renderGoal() {
+        const goalState = this.curGoalState;
+        if (!goalState) return '';
+        return `<div id="goal"><h1>Tactic State</h1><pre>${goalState}</pre></div>`;
     }
 
-    provideTextDocumentContent(uri: Uri, token: CancellationToken): string {
-        const uriString = uri.toString();
-        if (!this.documents.has(uriString)) {
-            if (this.leanInfoUrl.toString() !== uriString) {
-                throw new Error(`unsupported uri: ${uri}`);
+    private renderMessages() {
+        if (!this.curFileName) return ``;
+        return this.curMessages.map((m) => {
+            const i = m.text.indexOf("\n");
+            if (i == -1) {
+                return `<div class="message ${m.severity}"><h1>${m.severity}: ${m.caption}</h1><pre>${m.text}</pre></div>`
             }
-            const doc = new InfoDocument(this.server, this.leanDocs);
-            doc.onChanged(() => this.changedEmitter.fire(uri));
-            this.documents.set(uriString, doc);
-        }
-        return this.documents.get(uriString).text;
+            return `<div class="message ${m.severity}">
+                <h1>${m.severity}: ${m.caption} ${m.text.substring(0, i)}</h1>
```

Comment:
> So it draws a lot of attention

I'm not sure, the light red text of the caption seems a bit lighter than the black text; at least with my settings:
![vscode-faint-red](https://user-images.githubusercontent.com/313929/27189488-cae6154e-51f1-11e7-9224-01834b694e15.png)
It also puts the first line away from the main text, indented, and in a different color, which is something we don't do anywhere else.  It actually makes it kind of hard to read for me, but maybe I'd get used to it.

What I'm mainly concerned about is when text gets split between the caption and the pre block:
![vscode-split-text](https://user-images.githubusercontent.com/313929/27189592-23265e94-51f2-11e7-809c-6821ec17e15b.png)

> Maybe this is something which should change in Lean?

I have actually thought about the opposite: removing captions altogether.  We don't show them in emacs, or in the vscode diagnostics.

Personally, I think that we should not try to have special captions, or even guess that the first line could be used as a caption.  If we really want to highlight portions of a message, then we should use structured markup (server-side and also transmitted over the server interface): this would also allow us to tag the parts of a message which should be syntax highlighted as lean code.

For now, I'd strongly prefer that we do not try to guess any captions--this breaks very easily.  We can add more specific captions on the server side, if you think that's a good idea.